tests-and-examples/tests/cases/credential.rs

```diff
@@ -0,0 +1,242 @@
+            return Ok(None);
+        }
+        let new_cred = ServerCredentialsBuilder::new()
+            .add_cert(PD_CRT.into(), PD_KEY.into())
+            .root_cert(CA_CRT, CertificateRequestType::DontRequestClientCertificate);
+        self.reload = true;
+        Ok(Some(new_cred))
+    }
+}
+
+struct DataMeetFail {}
+
+impl UsrFetcher for DataMeetFail {
+    fn fetch(&mut self) -> Result<Option<ServerCredentialsBuilder>, Box<dyn std::error::Error>> {
+        let mut f = fs::File::open("Forsaken/Land")?;
+        let mut content = String::new();
+        f.read_to_string(&mut content)?;
+        Ok(None)
+    }
+}
+
+#[test]
+fn test_reload_new() {
+    let env = Arc::new(EnvBuilder::new().build());
+    let cred = ServerCredentialsBuilder::new()
+        .add_cert(SERVER_CRT.into(), SERVER_KEY.into())
+        .root_cert(CA_CRT, CertificateRequestType::DontRequestClientCertificate)
+        .build_with_fetcher(Box::new(DataReloadNew { reload: false }));
+    let service = create_greeter(GreeterService);
+    let mut server = ServerBuilder::new(env.clone())
+        .register_service(service)
+        .bind_secure("localhost", 0, cred)
+        .build()
+        .unwrap();
+    server.start();
+
+    let port = server.bind_addrs()[0].1;
+    let cred = ChannelCredentialsBuilder::new()
+        .root_cert(CA_CRT.into())
+        .build();
+    let ch = ChannelBuilder::new(env).secure_connect(&format!("localhost:{}", port), cred);
+    let client = GreeterClient::new(ch);
+
+    for _ in 0..10 {
+        let mut req = HelloRequest::default();
+        req.set_name("world".to_owned());
+        let reply = client.say_hello(&req).expect("rpc");
+        assert_eq!(reply.get_message(), "Hello world");
+    }
+}
+
+#[test]
+fn test_reload_fail_open() {
+    let env = Arc::new(EnvBuilder::new().build());
+    let cred = ServerCredentialsBuilder::new()
+        .add_cert(SERVER_CRT.into(), SERVER_KEY.into())
+        .root_cert(CA_CRT, CertificateRequestType::DontRequestClientCertificate)
+        .build_with_fetcher(Box::new(DataMeetFail {}));
+    let service = create_greeter(GreeterService);
+    let mut server = ServerBuilder::new(env.clone())
+        .register_service(service)
+        .bind_secure("localhost", 0, cred)
+        .build()
+        .unwrap();
+    server.start();
+
+    let port = server.bind_addrs()[0].1;
+    let cred = ChannelCredentialsBuilder::new()
+        .root_cert(CA_CRT.into())
+        .build();
+    let ch = ChannelBuilder::new(env).secure_connect(&format!("localhost:{}", port), cred);
+    let client = GreeterClient::new(ch);
+
+    for _ in 0..10 {
+        let mut req = HelloRequest::default();
+        req.set_name("world".to_owned());
+        let reply = client.say_hello(&req).expect("rpc");
+        assert_eq!(reply.get_message(), "Hello world");
+    }
+}
+
+#[allow(dead_code)]
+static CA_CRT: &str = "-----BEGIN CERTIFICATE-----
+MIIDRTCCAi2gAwIBAgIUGrzcBxX5bVapB86XYGPetE10wiEwDQYJKoZIhvcNAQEL
```

Comment:
Could you move them into files? Large string literal increases compile time.

Reply:
Fix